src/vs/base/common/trackedDisposable.ts

```diff
@@ -3,12 +3,38 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import { Emitter } from './event.js';
 import { Disposable } from '../../base/common/lifecycle.js';
 
 /**
- * Disposable class that tracks its own {@linkcode disposed} state.
+ * Disposable object that tracks its {@linkcode disposed} state
+ * as a public attribute and provides the {@linkcode onDispose}
+ * event to subscribe to.
  */
-export class TrackedDisposable extends Disposable {
+export abstract class TrackedDisposable extends Disposable {
+	/**
+	 * Private emitter for the `onDispose` event.
+	 */
+	private readonly _onDispose = this._register(new Emitter<void>());
+
+	/**
+	 * The event is fired when this object is disposed.
+	 * Note! Executes the callback immediately if already disposed.
+	 *
+	 * @param callback The callback function to be called on updates.
+	 */
+	public onDispose(callback: () => void): this {
```

Comment:
Why not an Event?

Reply:
When you add new common helpers that might be generally useful to people, you can mention them in `#dev-discuss` so that everyone knows they exist and can use them. Also so we can talk about the design if it's a general purpose thing that might get widely adopted.

Reply:
Will start a thread for this 👍 I find myself often need this logic on my classes and wondering why we don't have it on the base `Disposable` class 🤔